src/auth/rest.rs

```diff
@@ -0,0 +1,210 @@
+        self.username_placeholder = s;
+        self
+    }
+
+    /// specify the placeholder string in the rest of the fields that would be replaced by the password
+    pub fn with_password_placeholder(mut self, s: String) -> Self {
+        self.password_placeholder = s;
+        self
+    }
+
+    /// specify HTTP method
+    pub fn with_method(mut self, s: http::Method) -> Self {
+        self.method = s;
+        self
+    }
+
+    /// specify HTTP url
+    pub fn with_url(mut self, s: String) -> Self {
+        self.url = s;
+        self
+    }
+
+    /// specify HTTP body (ignored if does not apply for method)
+    pub fn with_body(mut self, s: String) -> Self {
+        self.body = s;
+        self
+    }
+
+    /// specify JSON selector to be used to extract the value from the response
+    /// format is serde_json's Value.pointer()
+    pub fn with_selector(mut self, s: String) -> Self {
+        self.selector = s;
+        self
+    }
+
+    /// specify the value the json selector's result should match to
+    pub fn with_regex(mut self, s: String) -> Self {
+        self.regex = s;
+        self
+    }
+
+    ///
+    pub fn build(self) -> RestAuthenticator {
+        RestAuthenticator {
+            username_placeholder: self.username_placeholder,
+            password_placeholder: self.password_placeholder,
+            method: self.method,
+            url: self.url,
+            body: self.body,
+            selector: self.selector,
+            regex: Regex::new(&self.regex).unwrap(),
+        }
+    }
+}
+
+impl RestAuthenticator {
+    fn fill_encoded_placeholders(&self, string: &str, username: &str, password: &str) -> String {
+        string
+            .replace(&self.username_placeholder, username)
+            .replace(&self.password_placeholder, password)
+    }
+}
+
+impl Authenticator for RestAuthenticator {
+    fn authenticate(
+        &self,
+        _username: &str,
+        _password: &str,
+    ) -> Box<Future<Item = bool, Error = ()> + Send> {
+
+        let username_url =
+            utf8_percent_encode(_username, PATH_SEGMENT_ENCODE_SET).collect::<String>();
+        let password_url =
+            utf8_percent_encode(_password, PATH_SEGMENT_ENCODE_SET).collect::<String>();
+        let url = self.fill_encoded_placeholders(&self.url, &username_url, &password_url);
+
+        let username_json = encode_string_json(_username);
+        let password_json = encode_string_json(_password);
+        let body = self.fill_encoded_placeholders(&self.body, &username_json, &password_json);
+
+        // FIXME: need to clone too much, just to keep tokio::spawn() happy, with its 'static requirement. is there a way maybe to work this around with proper lifetime specifiers? Or is it better to just clone the whole object?
+        let method = self.method.clone();
+        let selector = self.selector.clone();
+        let regex = self.regex.clone();
+
+        //        println!("{} {}", url, body);
```

Comment:
Pehaps we should just remove all the debugging by println :-). Consider replacing with `debug!`

Reply:
I've replaced it with `debug!`